genmc: Bug fixes

michaliskok opened the pull request:
This PR bumps the GenMC version to 0.16.1:
 - Fixes the incorrect treatment of `atomic_fetch_{umax,umin}` (rust-lang/miri#4572)
 - Fixes overflowing on large allocations (rust-lang/miri#4572)
 - Collects all wrappers around the GenMC API in a single file (`Exploration.cpp`)
 - Performs minor cleanups
 - Fixes a couple of... FIXMEs

r? @RalfJung
(Each commit includes a description and can be reviewed separately.)

RalfJung commented on src/concurrency/genmc/mod.rs:
```diff
@@ -622,15 +622,15 @@ impl GenmcCtx {
+            .handle_free(self.active_thread_genmc_tid(machine), address.bytes());
+        if let Some(error) = free_result.as_ref() {
             // FIXME(genmc): improve error handling.
             // An error was detected, so we get the error string from GenMC.
```
"so we get the error string from GenMC" isn't the case any more, so the last line can be removed.

(Github does not support code suggestions for commit-by-commit review. :(  )

RalfJung reviewed (commented):
Thanks!

The last commit mentions issues with memory allocation; what exactly are you referring to?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/miri/4733/ea0d7001155fe2b822fd03eb39b6fca94d4d648e..4cbc75e9fdb628a24793172dfc878d8799e26827)*

RalfJung commented:
> The last commit mentions issues with memory allocation; what exactly are you referring to?

Ah, the PR description mentions "Fixes overflowing on large allocations". Is this about a single large allocation, or many allocations accumulating to exhaust the 4GB per-thread limit? How large is "large"? Is it possible to test for that? What happens now when the per-thread 4GB memory region is exhausted?

RalfJung commented on genmc-sys/cpp/include/MiriInterface.hpp:
```diff
@@ -126,7 +126,7 @@ struct MiriGenmcShim : private GenMCDriver {
     /**** Memory (de)allocation ****/
     auto handle_malloc(ThreadId thread_id, uint64_t size, uint64_t alignment) -> uint64_t;
-    auto handle_free(ThreadId thread_id, uint64_t address) -> bool;
+    auto handle_free(ThreadId thread_id, uint64_t address) -> std::unique_ptr<std::string>;
```
> It's used in [handle_execution_end](https://github.com/rust-lang/miri/blob/59009fd7f18228439fa5623936478873d1c3a1c0/genmc-sys/src/lib.rs#L383).

Ah, fair.

No need to add a new type, but please add a short comment here saying that a non-null return value indicates an error.

RalfJung replied:
And also add such a comment to `handle_execution_end`, please.

RalfJung commented:
> The new commit makes GenMC crash if the limit is exceeded, 

That may make it tricky to add a test then, depending on the exit status this will trigger.

michaliskok commented:
> > The new commit makes GenMC crash if the limit is exceeded,
> 
> That may make it tricky to add a test then, depending on the exit status this will trigger.

The alternative would be to return an error code to Miri and let Miri crash.

RalfJung commented on tests/genmc/fail/simple/alloc_large.rs:
```diff
@@ -0,0 +1,25 @@
+//@compile-flags: -Zmiri-genmc -Zmiri-disable-stacked-borrows -Zmiri-ignore-leaks
+//@error-in-other-file: resource exhaustion
+
+// Ensure that we panic if GenMC fails to fulfill an allocation.
```
```suggestion
// Ensure that we emit a proper error if GenMC fails to fulfill an allocation.
```

michaliskok commented:
> I tried what happened when one thread allocates more than 4GB memory in total (but not at the same time, i.e. we allocate chunks of 512MB and free them again):
> 
> ```
> ./miri run --features=genmc bench-cargo-miri/big-allocs/src/main.rs -Zmiri-genmc
> ```
> 
> With this PR, that leads to a panic:
> 
> ```
> thread 'rustc' (123993) panicked at src/concurrency/genmc/mod.rs:597:9:
> assertion `left != right` failed: GenMC malloc returned nullptr.
> ```
> 
> EDIT: I get the same with an 8GB allocation, using this test:
> 
> ```rust
> fn main() {
>     let _v = Vec::<u8>::with_capacity(8 * 1024 * 1024 * 1024);
> }
> ```
> 
> So, I can't reproduce the GenMC crash.

Indeed, sorry about that.

I thought I made GenMC panic in all cases, but that's not the case. For allocations taking place via `handleMalloc`, GenMC returns nullptr (and potentially an error trace, but error reporting is not hooked up yet anyway). I added a test ensuring that Miri produces an error report if addresses get depleted.

I also incorporated your other comments.

RalfJung commented on genmc-sys/cpp/include/MiriInterface.hpp:
```diff
@@ -125,8 +125,11 @@ struct MiriGenmcShim : private GenMCDriver {
     auto handle_malloc(ThreadId thread_id, uint64_t size, uint64_t alignment) -> uint64_t;
-    auto handle_free(ThreadId thread_id, uint64_t address) -> bool;
+
+    /** Returns null or an error string if an error occurs. */
```
I had edited this suggestion after initially posting it, looks like that edit got lost?

michaliskok commented:
@rustbot ready
